action/rlp_tx.go

```diff
@@ -82,35 +81,95 @@ func reconstructSignedRlpTxFromSig(tx rlpTransaction, chainID uint32, sig []byte
+func DecodeRawTx(rawData string, chainID uint32) (*types.Transaction, bool, error) {
 	//remove Hex prefix and decode string to byte
-	rawData = strings.Replace(rawData, "0x", "", -1)
-	rawData = strings.Replace(rawData, "0X", "", -1)
-	var dataInString []byte
-	dataInString, err = hex.DecodeString(rawData)
+	dataInString, err := hex.DecodeString(util.Remove0xPrefix(rawData))
 	if err != nil {
-		return
+		return nil, false, err
 	}
 
 	// decode raw data into rlp tx
-	tx = &types.Transaction{}
-	err = rlp.DecodeBytes(dataInString, tx)
-	if err != nil {
-		return
+	unwrapper := &transactionUnwrapper{}
+	if err = rlp.DecodeBytes(dataInString, unwrapper); err != nil {
+		return nil, false, err
 	}
+	return unwrapper.tx, unwrapper.isEthEncoding, nil
+}
 
-	// extract signature and recover pubkey
+// GetSignatureFromEthTX calculates signature from tx
+func GetSignatureFromEthTX(tx *types.Transaction, chainID uint32) ([]byte, error) {
+	if tx == nil {
+		return nil, errors.New("pointer is nil")
+	}
 	v, r, s := tx.RawSignatureValues()
 	recID := uint32(v.Int64()) - 2*chainID - 8
-	sig = make([]byte, 64, 65)
+	sig := make([]byte, 64, 65)
 	rSize := len(r.Bytes())
 	copy(sig[32-rSize:32], r.Bytes())
 	sSize := len(s.Bytes())
 	copy(sig[64-sSize:], s.Bytes())
 	sig = append(sig, byte(recID))
+	return sig, nil
+}
+
+type (
+	// transactionUnwrapper is a unwrapper for the Ethereum transaction.
+	transactionUnwrapper struct {
+		tx            *types.Transaction
+		isEthEncoding bool
+	}
 
-	// recover public key
-	rawHash := types.NewEIP155Signer(big.NewInt(int64(chainID))).Hash(tx)
-	pubkey, err = crypto.RecoverPubkey(rawHash[:], sig)
-	return
+	legacyTxWithEncodingType struct {
+		// data fields of legacy transaction.
+		Nonce    uint64
+		GasPrice *big.Int
+		Gas      uint64
+		To       *common.Address `rlp:"nil"`
+		Value    *big.Int
+		Data     []byte
+		V, R, S  *big.Int
+		// extra field to identify tx signed from ledgers which use native signing method
+		EncodingType []byte `rlp:"optional"`
+	}
+)
+
+// DecodeRLP implements eth.rlp.Decoder
+func (unwrapper *transactionUnwrapper) DecodeRLP(s *rlp.Stream) error {
+	kind, _, err := s.Kind()
+	if err != nil {
+		return err
+	}
+	switch kind {
+	case rlp.List:
+		outter := legacyTxWithEncodingType{}
+		if err := s.Decode(&outter); err != nil {
+			return err
+		}
+		unwrapper.tx = types.NewTx(outter.exportLegacyTx())
+		unwrapper.isEthEncoding = outter.isEthEncoding()
+		return nil
+	default:
+		return rlp.ErrExpectedList
+	}
+}
+
+func (tx *legacyTxWithEncodingType) exportLegacyTx() *types.LegacyTx {
+	return &types.LegacyTx{
+		Nonce:    tx.Nonce,
+		GasPrice: tx.GasPrice,
+		Gas:      tx.Gas,
+		To:       tx.To,
+		Value:    tx.Value,
+		Data:     tx.Data,
+		V:        tx.V,
+		R:        tx.R,
+		S:        tx.S,
+	}
+}
+
+func (tx *legacyTxWithEncodingType) isEthEncoding() bool {
+	if len(tx.EncodingType) > 0 && hex.EncodeToString(tx.EncodingType) == "01" {
```

Comment:
pending Leo's final implementation, i think we discussed using a 4-byte magic number